File: 5-two-pointers/4-variable-size-sliding-window/main.py
# Variable Size Sliding Window - Python Example

def min_subarray_len(target, arr):
    start = 0
    curr_sum = 0
    min_len = float('inf')
    
    for end in range(len(arr)):
        curr_sum += arr[end]
        while curr_sum >= target:
            min_len = min(min_len, end - start + 1)
            curr_sum -= arr[start]
            start += 1

    return min_len if min_len != float('inf') else 0

arr = [2,3,1,2,4,3]
target = 7
print("Minimum subarray length with sum >= ", target, ":", min_subarray_len(target, arr))

File: 5-two-pointers/4-variable-size-sliding-window/main.cpp
// Variable Size Sliding Window - C++ Example

#include <iostream>
#include <vector>
#include <climits>

int minSubarrayLen(int target, const std::vector<int>& arr) {
    int start = 0, currSum = 0, minLen = INT_MAX;

    for (int end = 0; end < arr.size(); ++end) {
        currSum += arr[end];
        while (currSum >= target) {
            minLen = std::min(minLen, end - start + 1);
            currSum -= arr[start++];
        }
    }

    return minLen == INT_MAX ? 0 : minLen;
}

int main() {
    std::vector<int> arr = {2,3,1,2,4,3};
    int target = 7;
    std::cout << "Minimum subarray length with sum >= " << target << ": " << minSubarrayLen(target, arr) << std::endl;
    return 0;
}
